core/src/main/java/org/apache/iceberg/stats/StatsReader.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.iceberg.relocated.com.google.common.io.ByteStreams;
+
+public class StatsReader implements Closeable {
+
+  private final long fileLength;
+  private final SeekableInputStream input;
+  private Optional<Integer> knownFooterSize;
+  private Optional<FileMetadata> knownFileMetadata = Optional.empty();
+
+  public StatsReader(InputFile inputFile, Optional<Integer> footerSize) {
+    this.fileLength = inputFile.getLength();
+    this.input = Preconditions.checkNotNull(inputFile, "inputFile is null").newStream();
+    this.knownFooterSize = Preconditions.checkNotNull(footerSize, "footerSize is null");
+    footerSize.ifPresent(size ->
+        Preconditions.checkArgument(0 < size && size <= fileLength - StatsFormat.getMagic().length,
+            "Invalid footer size: %s", size));
+  }
+
+  public FileMetadata getFileMetadata() throws IOException {
+    if (!knownFileMetadata.isPresent()) {
+      int footerSize = getFooterSize();
+      byte[] footer = readInput(fileLength - footerSize, footerSize);
+
+      checkMagic(footer, 0);
+      checkMagic(footer, footerSize - 4);
+
+      int fileFormatVersion = StatsFormat.readIntegerLittleEndian(footer, footerSize - 8);
+      Preconditions.checkState(
+          fileFormatVersion == StatsFormat.CURRENT_FORMAT_VERSION,
+          "Unsupported format version %s, %s is the latest supported",
+          fileFormatVersion,
+          StatsFormat.CURRENT_FORMAT_VERSION);
+
+      int flags = StatsFormat.readIntegerLittleEndian(footer, footerSize - 12);
+      Preconditions.checkState((flags & ~StatsFormat.SUPPORTED_FLAGS) == 0, "Unsupported flags: %s", flags);
+      boolean compressed = isFlagSet(flags, StatsFormat.FLAG_COMPRESSED);
+
+      int reserved = StatsFormat.readIntegerLittleEndian(footer, footerSize - 16);
+      Preconditions.checkState(reserved == 0, "Unexpected reserved bytes value: %s", reserved);
+
+      int footerPayloadSize = StatsFormat.readIntegerLittleEndian(footer, footerSize - 20);
+      Preconditions.checkState(footerPayloadSize == footerSize - 24,
+          "Unexpected footer payload size value %s for footer size %s", footerPayloadSize, footerSize);
+
+      FileMetadata readFileMetadata;
+      ByteBuffer footerPayload = ByteBuffer.wrap(footer, 4, footerPayloadSize);
+      if (!compressed) {
+        readFileMetadata = parseFileMetadata(footerPayload);
+      } else {
+        ByteBuffer footerJson = StatsFormat.decompressFooterPayload(footerPayload);
+        readFileMetadata = parseFileMetadata(footerJson);
+      }
+      this.knownFileMetadata = Optional.of(readFileMetadata);
+    }
+    return knownFileMetadata.get();
+  }
+
+  /**
+   * @throws IOException          when I/O error occurs
+   * @throws UncheckedIOException when I/O error occurs
+   */
+  // Note: The method is marked as throwing IOException to allow future implementation evolution.
+  // Currently, the exception isn't being thrown.
```

Comment:
In Iceberg, we typically omit `IOException` and wrap with `UncheckedIOException` because that has to happen in so many places anyway. I think you can remove the exception.